### PR TITLE
Implement and test the following flags for podman run and create

### PR DESCRIPTION
**Pull request by umohnani8**
memory, memory-reservation, memory-swap, memory-swappiness, kernel-memory,
cpu-period, cou-quota, cpu-shares, cpus, cpuset-cpus, cpuset-mems,
blkio-weight, blkio-weight-device, sysctl, and ulimit

Signed-off-by: umohnani8 <umohnani@redhat.com>

### CODE REVIEW

File: cmd/podman/spec.go [hunk 1]
```diff
@@ -210,10 +210,8 @@ func createConfigToOCISpec(config *createConfig) (*spec.Spec, error) {
 	if config.Hostname != "" {
 		g.AddProcessEnv("HOSTNAME", config.Hostname)
 	}
-
-	for _, sysctl := range config.Sysctl {
-		s := strings.SplitN(sysctl, "=", 2)
-		g.AddLinuxSysctl(s[0], s[1])
+	for sysctlKey, sysctlVal := range config.Sysctl {
+		g.AddLinuxSysctl(sysctlKey, sysctlVal)
```
Comment (rhatdan) [inline cmd/podman/spec.go]: Are you verifying the syscalls as legal?  We can do that in a separate PR.

File: cmd/podman/spec.go [hunk 2]
```diff
@@ -246,14 +243,18 @@ func createConfigToOCISpec(config *createConfig) (*spec.Spec, error) {
 	if config.Resources.CPUPeriod != 0 {
 		g.SetLinuxResourcesCPUPeriod(config.Resources.CPUPeriod)
 	}
+	if config.Resources.CPUs != 0 {
+		g.SetLinuxResourcesCPUPeriod(100000)
```
Comment (rhatdan) [inline cmd/podman/spec.go]: Should this be a constant?

File: cmd/podman/spec.go [hunk 3]
```diff
@@ -356,6 +357,13 @@ func createConfigToOCISpec(config *createConfig) (*spec.Spec, error) {
 		return nil, err
 	}
 
+	// BLOCK IO
+	blkio, err := config.CreateBlockIO()
```
Comment (rhatdan) [inline cmd/podman/spec.go]: Does this happen no matter what?

File: cmd/podman/spec.go (hunk 2 above)
Comment (mheon) [reply cmd/podman/spec.go]: Agree, should probably be a constant

File: cmd/podman/spec.go (hunk 1 above)
Comment (umohnani8) [reply cmd/podman/spec.go]: Legal as in if they are key=value pairs? If so, yes it i being checked in create.go

File: cmd/podman/spec.go (hunk 2 above)
Comment (umohnani8) [reply cmd/podman/spec.go]: fixed

File: cmd/podman/spec.go (hunk 3 above)
Comment (umohnani8) [reply cmd/podman/spec.go]: I added a check after to see if `blkio` is nil or not, and if its not it is added to the spec. Is that better @rhatdan ? Because a bunch of flags are used in the `CreateBlockIO` function, which would be a huge if statement.

File: cmd/podman/spec.go (hunk 1 above)
Comment (mheon) [reply cmd/podman/spec.go]: Only certain sysctls are valid in containers. Not sure which. Docker likely maintains a list.

Comment (rhatdan) [reply cmd/podman/spec.go]:
```
// ValidateSysctl validates a sysctl and returns it.
func ValidateSysctl(val string) (string, error) {
        validSysctlMap := map[string]bool{
                "kernel.msgmax":          true,
                "kernel.msgmnb":          true,
                "kernel.msgmni":          true,
                "kernel.sem":             true,
                "kernel.shmall":          true,
                "kernel.shmmax":          true,
                "kernel.shmmni":          true,
                "kernel.shm_rmid_forced": true,
        }
        validSysctlPrefixes := []string{
                "net.",
                "fs.mqueue.",
        }
        arr := strings.Split(val, "=")
        if len(arr) < 2 {
                return "", fmt.Errorf("sysctl '%s' is not whitelisted", val)
        }
        if validSysctlMap[arr[0]] {
                return val, nil
        }

        for _, vp := range validSysctlPrefixes {
                if strings.HasPrefix(arr[0], vp) {
                        return val, nil
                }
        }
        return "", fmt.Errorf("sysctl '%s' is not whitelisted", val)
}
```

File: cmd/podman/spec.go (hunk 3 above)
Comment (rhatdan) [reply cmd/podman/spec.go]: SGTM

**Comment by mheon** [comment]
bot, retest this please

File: cmd/podman/spec.go (hunk 1 above)
Comment (umohnani8) [reply cmd/podman/spec.go]: Added validation.

File: cmd/podman/create_cli.go [hunk 4]
```diff
@@ -36,6 +36,48 @@ func convertStringSliceToMap(strSlice []string, delimiter string) (map[string]st
...
+		if validSysctlMap[arr[0]] {
+			sysctl[arr[0]] = arr[1]
+			continue
+		}
+
+		for _, prefix := range validSysctlPrefixes {
+			if strings.HasPrefix(arr[0], prefix) {
+				sysctl[arr[0]] = arr[1]
+				foundMatch = true
+			}
```
Comment (rhatdan) [inline cmd/podman/create_cli.go]: Could add a break here to get minimal speedup.

**Comment by rhatdan** [comment]
LGTM other then an incredibly small NIT.

File: cmd/podman/create_cli.go (hunk 4 above)
Comment (umohnani8) [reply cmd/podman/create_cli.go]: fixed

**Comment by mheon** [comment]
LGTM

**Comment by mheon** [comment]
@rh-atomic-bot r+
